### ISSUE TITLE
Reproducibility problem with optuna.integration.OptunaSearchCV

### ISSUE THREAD

**Issue by cconstantinou73** [optuna/optuna#2540]
## I expect to have reproducible results

- Optuna version: '2.0.0'
- Python version: Python 3.8.8
- OS: PRETTY_NAME="Ubuntu 18.04.3 LTS", VERSION_ID="18.04"
- (Optional) Other libraries and their versions:

## Error messages, stack traces, or logs

```
# error messages, stack traces, or logs
```

## Steps to reproduce

1. see script below
2.
3.

## Reproducible examples (optional)

```python

import optuna
import xgboost as xgb
from sklearn.datasets import load_boston

X, y = load_boston(return_X_y=True)

seed = 7
n_iter = 3
optimize = 'r2'
tuner_verbose = 0
n_jobs = -1
early_stopping_max_iters = 10
prune = False
fold = 3

regressor = xgb.XGBRegressor()

sampler = optuna.samplers.TPESampler(seed=seed)
study =  optuna.create_study(
            direction="maximize", sampler=sampler
        )

param_grid = {'eta':optuna.distributions.LogUniformDistribution(high=0.5, low=1e-06),
              'depth':optuna.distributions.IntUniformDistribution(high=11, low=1, step=1),
              'n_estimators':optuna.distributions.IntUniformDistribution(high=300, low=10, step=1),
              'random_strength':optuna.distributions.UniformDistribution(high=0.8, low=0),
              'l2_leaf_reg':optuna.distributions.IntLogUniformDistribution(high=200, low=1, step=1)
             }


model_grid = optuna.integration.OptunaSearchCV(
    estimator=regressor,
    param_distributions=param_grid,
    cv=fold,
    enable_pruning=prune,
    max_iter=early_stopping_max_iters,
    n_jobs=n_jobs,
    n_trials=n_iter,
    random_state=seed,
    scoring=optimize,
    study=study,
    refit=False,
    verbose=tuner_verbose,
    error_score="raise",
)

model_grid.fit(X, y)

print(model_grid.best_params_)

```

## Additional context (optional)

After trial and error i have noticed that if i comment out the n_jobs, study and random_state arguments and use numpy.random.seed(7) i get reproducible results but its desired to use n_jobs=-1 rather than the default value of n_jobs=1

**Comment by nzw0301** [comment]
I think that this is expected behaviour according to [the document](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.samplers.TPESampler.html#optuna.samplers.TPESampler.reseed_rng).

**Comment by cconstantinou73** [comment]
@nzw0301 I am not sure what you mean. In the above code example what is being done wrong? For example i want to run the script above more than once and i want to replicate the results of model_grid.best_params_. When i run the code above i get different results for model_grid.best_params_ after each run.

**Comment by nzw0301** [comment]
The code seems great, but we cannot get reproducible results due to the optuna's implementation for `n_jobs>1` or `n_jobs=-1`.  When we use `n_jobs>1`, optuna internally re-sest a different random seed to each sampler per job via `reseed_rng` method. Unfortunately, we cannot pass a seed value to this internal process.

**Comment by cconstantinou73** [comment]
@nzw0301 Thank you for your response. I was mainly interested for reproducible results when n_jobs>1. It must be noted though with the above implementation even if you set n_jobs=1 you still don't get reproducible results. As i already said before i only could get reproducible results if i do not specify neither the n_jobs (which defaults to 1) neither the study and neither the random_state and instead use np.random.seed(any_seed_number) before calling the optuna.integration.OptunaSearchCV, so not sure how specifying the study and random_state in the arguments does not result in reproducible results

**Comment by nzw0301** [comment]
It's my pleasure.

I'm not sure that I could understand when the code is reproducible. When I run the following code, the results seem fine. Could you give me an example?

```python
import optuna
import xgboost as xgb
from sklearn.datasets import load_boston

X, y = load_boston(return_X_y=True)

seed = 7
n_iter = 20
optimize = 'r2'
tuner_verbose = 0
n_jobs = 1
early_stopping_max_iters = 10
prune = False
fold = 3

res = []
for i in range(2):
    regressor = xgb.XGBRegressor()

    sampler = optuna.samplers.TPESampler(seed=seed)
    study = optuna.create_study(
    direction="maximize", sampler=sampler
    )

    param_grid = {'eta':optuna.distributions.LogUniformDistribution(high=0.5, low=1e-06),
    'depth':optuna.distributions.IntUniformDistribution(high=11, low=1, step=1),
    'n_estimators':optuna.distributions.IntUniformDistribution(high=300, low=10, step=1),
    'random_strength':optuna.distributions.UniformDistribution(high=0.8, low=0),
    'l2_leaf_reg':optuna.distributions.IntLogUniformDistribution(high=200, low=1, step=1)
    }

    model_grid = optuna.integration.OptunaSearchCV(
        estimator=regressor,
        param_distributions=param_grid,
        cv=fold,
        enable_pruning=prune,
        max_iter=early_stopping_max_iters,
        n_jobs=n_jobs,
        n_trials=n_iter,
        random_state=seed,
        scoring=optimize,
        study=study,
        refit=False,
        verbose=tuner_verbose,
        error_score="raise",
    )

    model_grid.fit(X, y)
    res.append(model_grid.best_params_)

res[0] == res[1]  # True
```

**Comment by cconstantinou73** [comment]
I cant seem to replicate the behaviour of inconsistency when n_jobs == 1 and defining a study and random_state, so i am assuming i might have done something wrong when using those before. In other words i get the expected behaviour of reproducibility when n_jobs == 1. Thank you for your responses @nzw0301

**Comment by nzw0301** [comment]
Awesome! I'm glad to hear that. Thanks to your feedback, I'm sending a pull request to improve the documentation of `n_jobs`.
